packages/extension-base/src/services/balance-service/transfer/ton-transfer.ts

```diff
@@ -0,0 +1,138 @@
+// Copyright 2019-2022 @subwallet/extension-base
+// SPDX-License-Identifier: Apache-2.0
+
+import { _ChainAsset } from '@subwallet/chain-list/types';
+import { INIT_FEE_JETTON_TRANSFER, TON_OPCODES, WORKCHAIN } from '@subwallet/extension-base/services/balance-service/helpers/subscribe/ton/consts';
+import { estimateTonTxFee } from '@subwallet/extension-base/services/balance-service/helpers/subscribe/ton/utils';
+import { _TonApi } from '@subwallet/extension-base/services/chain-service/types';
+import { _getContractAddressOfToken, _isJettonToken, _isNativeToken } from '@subwallet/extension-base/services/chain-service/utils';
+import { keyring } from '@subwallet/ui-keyring';
+import { beginCell, fromNano, internal, MessageRelaxed, toNano } from '@ton/core';
+import { Address, JettonMaster, JettonWallet, WalletContractV4 } from '@ton/ton';
+
+interface TonTransactionConfigProps {
+  tokenInfo: _ChainAsset;
+  from: string,
+  to: string,
+  networkKey: string,
+  value: string,
+  transferAll: boolean,
+  tonApi: _TonApi
+}
+
+export interface TonTransactionConfig {
+  from: string,
+  to: string,
+  networkKey: string,
+  value: string,
+  messages: MessageRelaxed[]
+  estimateFee: string;
+  seqno: number
+}
+
+export async function createTonTransaction ({ from, networkKey, to, tokenInfo, tonApi, transferAll, value }: TonTransactionConfigProps): Promise<[TonTransactionConfig | null, string]> {
+  if (_isNativeToken(tokenInfo)) {
+    return createTonNativeTransaction({ from, to, networkKey, tokenInfo, value: value || '0', transferAll: transferAll, tonApi });
+  }
+
+  if (_isJettonToken(tokenInfo)) {
+    return createJettonTransaction({ from, to, networkKey, tokenInfo, value: value || '0', transferAll: transferAll, tonApi });
+  }
+
+  return [null, value];
+}
+
+async function createTonNativeTransaction ({ from, networkKey, to, tonApi, transferAll, value }: TonTransactionConfigProps): Promise<[TonTransactionConfig | null, string]> {
+  const keyPair = keyring.getPair(from);
+  const tonAddress = Address.parse(from);
+  const walletContract = WalletContractV4.create({ workchain: WORKCHAIN, publicKey: Buffer.from(keyPair.publicKey) });
+  const contract = tonApi.open(walletContract);
+  const seqno = await contract.getSeqno();
+
+  let transferValue: string | undefined;
+
+  const messages = [
+    internal({
+      to: to,
+      value: fromNano(value),
+      bounce: false // todo: check and update the send bounced logic
+    })
+  ];
+
+  const estimateFee = await estimateTonTxFee(tonApi, messages, walletContract);
+
+  if (transferAll) {
+    const balance = await tonApi.getBalance(tonAddress);
+
+    transferValue = (balance - estimateFee).toString();
```

Comment:
I think this should be calculated beforehand

Reply:
Max transfer value should be calculated before this step

Reply:
I've implemented the max transfer value before. But it's required the sent amount to exactly estimate the txFee. The fee increases a bit with logarithm of sent amount.

Reply:
-> Solution: Use FreeBalance as the sending amount to estimate fee